File: answers/leetcode/Edit Distance/Edit Distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int[][] distances = new int[word1.length() + 1][word2.length() + 1];
        for (int i = 0; i <= word1.length(); i++) {
            for (int j = 0; j <= word2.length(); j++) {
                int distance = 0;
                if (i == 0 || j == 0) {
                    distance = i + j;
                } else {
                    distance = distances[i - 1][j - 1] + (word1.charAt(i - 1) == word2.charAt(j - 1) ? 0 : 1);
                    distance = Math.min(distance, distances[i - 1][j] + 1);
                    distance = Math.min(distance, distances[i][j - 1] + 1);
                }
                distances[i][j] = distance;
            }
        }
        return distances[word1.length()][word2.length()];
    }
}

File: answers/leetcode/Edit Distance/Edit Distance.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """

        length1 = len(word1) + 1
        length2 = len(word2) + 1
        dp = [[1000000 for i in range(length2)] for j in range(length1)]
        for i in range(length1):
            dp[i][0] = i
        for i in range(length2):
            dp[0][i] = i
        for i in range(1, length1):
            for j in range(1, length2):
                dp[i][j] = min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + int(word1[i - 1] != word2[j - 1]))
        return dp[-1][-1]
